TrkHitReco/src/Chi2Clusterer.cc

```diff
@@ -0,0 +1,180 @@
+
+    auto transPred = [&BkgHits] (const int i) {return BkgHits[i].chidx_;};
+    for (auto& cluster: clusters) std::transform(cluster.hits().begin(),cluster.hits().end(),cluster.hits().begin(),transPred);
+
+    if(diag_>1) dump(clusters,BkgHits);
+
+  }
+
+
+  //----------------------------------------------------------------------------------------------------------------------
+  void Chi2Clusterer::initClustering(const ComboHitCollection& chcol, std::vector<Chi2BkgHit>& BkgHits)
+  {
+     float maxTime(0);
+     for (size_t ich=0;ich<chcol.size();++ich) {
+       if (testflag_ && (!chcol[ich].flag().hasAllProperties(sigmask_) || chcol[ich].flag().hasAnyProperty(bkgmask_))) continue;
+       BkgHits.emplace_back(Chi2BkgHit(ich));
+       maxTime = std::max(maxTime,chcol[ich].correctedTime());
+     }
+     tbin_ = (maxTime+1.0)/float(numBuckets_);
+
+     auto resPred = [&chcol](const Chi2BkgHit& x, const Chi2BkgHit& y) {return chcol[x.chidx_].wireRes() < chcol[y.chidx_].wireRes();};
+     std::sort(BkgHits.begin(),BkgHits.end(),resPred);
+  }
+
+
+  //----------------------------------------------------------------------------------------------------------------------
+  void Chi2Clusterer::doClustering(const ComboHitCollection& chcol, std::vector<BkgCluster>& clusters, std::vector<Chi2BkgHit>& BkgHits)
+  {
+    unsigned niter(0);
+    float startChi2(0);
+    float deltaChi2(0);
+    //Check average Chi2 change. This is set to middle point of hit and seed radii(dhit_+dseed_/2)^2
+    while ( deltaChi2 < chi2Cut_ && niter < maxNiter_ ) {
+      ++niter;
+      formClusters(chcol, clusters, BkgHits);
+
+      float aveChi2(0);
+      float sumChi2(0);
+      unsigned nofCl(0);
+      for (const auto& cluster: clusters) {
+        if(cluster.hits().size() > 1) {
+          sumChi2 += cluster.points().chisquared();
+          ++nofCl;
+        }
+      }
+      aveChi2 = sumChi2/nofCl;
+      if(niter == 1) startChi2 = aveChi2;
+      deltaChi2 = aveChi2 - startChi2;
+    }
+    if(diag_>0) std::cout<<"Nof_iter\t"<<niter<<"\tTotal_nof_clusters\t"<<clusters.size()<<"\tAverage_delta_chi2\t"<<deltaChi2<<"\n";
+
+  }
+
+  //-------------------------------------------------------------------------------------------------------------------
+  // ChiClusterer strategy;
+  // Create a new cluster if there is none. For new hits, check how much cluster chi2 is going to change when a new hit is added for all existing clusters.
+  // For the minimum change, if the change is within dhit_ parameter, add it to the cluster. Create a new cluster if change is greater than dseed_.
+  // Remove clusters with only single combo hit and try again to place them into existing clusters.
+  unsigned Chi2Clusterer::formClusters(const ComboHitCollection& chcol, std::vector<BkgCluster>& clusters, std::vector<Chi2BkgHit>& BkgHits)
+  {
+    std::array<std::vector<int>, numBuckets_> clusterIndices;
```

Comment:
This structure could make inefficient use of memory, since each vector will start out empty and get extended onto the heap in small increments.  do we know what the typical size of these vectors is?
Also, since the vector contents are used as an index, it would be better to code as size_t (negative values are impossible/illegal).

Reply:
This is remnant from TNTClusterer, we can discuss it further.